tests/integration/exec.bats

```diff
@@ -476,3 +476,19 @@ EOF
 	[ "$status" -eq 0 ]
 	[[ "${lines[0]}" == "Cpus_allowed_list:	$all_cpus" ]]
 }
+
+@test "runc exec with RLIMIT_NOFILE" {
+	update_config '.process.capabilities.bounding = ["CAP_SYS_RESOURCE"]'
+	update_config '.process.rlimits = [{"type": "RLIMIT_NOFILE", "hard": 65536, "soft": 65536}]'
+
+	runc run -d --console-socket "$CONSOLE_SOCKET" test_busybox
+	[ "$status" -eq 0 ]
+
+	update_config '.process.args = ["/bin/sh", "-c", "ulimit -n"]'
+	runc run test_ulimit
+	[[ "${output}" == "65536" ]]
```

Comment:
Intrestingly, I don't know how this PR impacts `runc run` with `RLIMIT_NOFILE` in `centos7`:
```
not ok 91 runc exec with RLIMIT_NOFILE
# (in test file tests/integration/exec.bats, line 489)
#   `[[ "${output}" == "65536" ]]' failed
#
# runc run -d --console-socket /tmp/bats-run-PQBrqc/runc.wbkYSh/tty/sock test_busybox (status=0):
#
# runc run test_ulimit (status=0):
# 1024
```

As it will not fail in #4267 .

@kolyshkin @cyphar @AkihiroSuda Do you know why?

Reply:
@lifubang first, there's no such code (of running a second container with modified `.process.args`) in #4267.

Second, you forgot to check the exit code (here and below).

Third, I think you don't have to add CAP_SYS_RESOURCE as this is the capability of the container itself, and rlimits are set by runc which already have it.

Finally, it looks like you just found a new bug -- `RLIMIT_NOFILE` is not being set in CentOS 7 for `runc run`. Can confirm this

Reply:
@lifubang the issue is, on CentOS 7 the hard limit for NOFILE is 4096. For some reason yet unknown to me, raising the limit is not possible for runc via prlimit(2) syscall, and it returns no error. Maybe it's a bug in CentOS 7 kernel (related to namespaces I guess).

Two ways to workaround it:
1. Use lower limits (soft <= 1024, hard <= 4096) in the test code.
2. Add something like this to the test:

```diff
function setup() {
+	prlimit --nofile=99999:99999 -p $$
	setup_busybox
}
```

Surely, I won't recommend item 2, but it works.

Reply:
Same in Ubuntu 20.04 LTS (kernel 5.15) and Fedora 40 (kernel 6.8) -- if the current hard limit is equal or lower than the one we're trying to set using prlimit syscall, it returns no error but the limit is not changed.

To repro this locally:
```console
$ sudo
# ulimit -H -n 65536
# bats tests/integration/ulimit.bats
root@kir-tp1:/home/kir/git/runc# bats tests/integration/ulimit.bats 
ulimit.bats
 ✗ runc run with RLIMIT_NOFILE
   (in test file tests/integration/ulimit.bats, line 19)
     `[[ "${output}" == "65536" ]]' failed
   runc spec (status=0):
   
   runc run test_ulimit (status=0):
   1024
 ✓ runc exec with RLIMIT_NOFILE
 ✗ runc run+exec two containers with RLIMIT_NOFILE
   (in test file tests/integration/ulimit.bats, line 45)
     `[[ "${output}" == "65536" ]]' failed
   runc spec (status=0):
   
   runc run -d --console-socket /tmp/bats-run-TSa3k7/runc.jz4uuS/tty/sock test_busybox (status=0):
   
   runc run test_ulimit (status=0):
   1024

3 tests, 2 failures
```

Here's the test code (`tests/integration/ulimit.bats`):
```bash
#!/usr/bin/env bats

load helpers

function setup() {
	setup_busybox
}

function teardown() {
	teardown_bundle
}

@test "runc run with RLIMIT_NOFILE" {
	update_config '.process.rlimits = [{"type": "RLIMIT_NOFILE", "hard": 65536, "soft": 65536}]'

	update_config '.process.args = ["/bin/sh", "-c", "ulimit -n"]'
	runc run test_ulimit
	[ "$status" -eq 0 ]
	[[ "${output}" == "65536" ]]
}

@test "runc exec with RLIMIT_NOFILE" {
	update_config '	 .process.rlimits = [{"type": "RLIMIT_NOFILE", "hard": 2000, "soft": 1000}]'

	runc run -d --console-socket "$CONSOLE_SOCKET" nofile
	[ "$status" -eq 0 ]

	for ((i = 0; i < 100; i++)); do
		runc exec nofile sh -c 'ulimit -n'
		echo "[$i] $output"
		[[ "${output}" == "1000" ]]
	done
}

@test "runc run+exec two containers with RLIMIT_NOFILE" {
	update_config '.process.capabilities.bounding = ["CAP_SYS_RESOURCE"]'
	update_config '.process.rlimits = [{"type": "RLIMIT_NOFILE", "hard": 65536, "soft": 65536}]'

	runc run -d --console-socket "$CONSOLE_SOCKET" test_busybox
	[ "$status" -eq 0 ]

	update_config '.process.args = ["/bin/sh", "-c", "ulimit -n"]'
	runc run test_ulimit
	[ "$status" -eq 0 ]
	[[ "${output}" == "65536" ]]

	# issue: https://github.com/opencontainers/runc/issues/4195
	runc exec test_busybox /bin/sh -c "ulimit -n"
	[ "$status" -eq 0 ]
	[[ "${output}" == "65536" ]]
}
```
If you set hard ulimit -n to 65537, it works again.

Reply:
Here's another addition: we do have a test case for setting RLIMIT_NOFILE during runc exec, but the test case is masking the issue because it sets Cur and Max to the same value (and so Go runtime does not do any tricks with saving/restoring NOFILE).

With this change, the test fails:
```diff
diff --git a/libcontainer/integration/exec_test.go b/libcontainer/integration/exec_test.go
index 9dd056ad..f4b8b839 100644
--- a/libcontainer/integration/exec_test.go
+++ b/libcontainer/integration/exec_test.go
@@ -140,7 +140,7 @@ func testRlimit(t *testing.T, userns bool) {
        // the Setrlimit call happens early enough that we still have permissions to raise the limit.
        ok(t, unix.Setrlimit(unix.RLIMIT_NOFILE, &unix.Rlimit{
                Max: 1024,
-               Cur: 1024,
+               Cur: 512,
        }))
 
        out := runContainerOk(t, config, "/bin/sh", "-c", "ulimit -n")
```

With the above change, the test now fails like this:

```console
[kir@kir-tp1 runc]$ go1.20.14 test -exec sudo  -v libcontainer/integration -run Rlimit
package libcontainer/integration is not in GOROOT (/home/kir/sdk/go1.20.14/src/libcontainer/integration)
[kir@kir-tp1 runc]$ go1.20.14 test -exec sudo  -v ./libcontainer/integration -run Rlimit
=== RUN   TestRlimit
    exec_test.go:148: expected rlimit to be 1025, got 512
--- FAIL: TestRlimit (0.11s)
=== RUN   TestUsernsRlimit
    exec_test.go:148: expected rlimit to be 1025, got 512
--- FAIL: TestUsernsRlimit (0.10s)
=== RUN   TestExecInUsernsRlimit
    execin_test.go:129: expected rlimit to be 1026, got 512
--- FAIL: TestExecInUsernsRlimit (0.14s)
=== RUN   TestExecInRlimit
    execin_test.go:129: expected rlimit to be 1026, got 512
--- FAIL: TestExecInRlimit (0.15s)
FAIL
FAIL	github.com/opencontainers/runc/libcontainer/integration	0.542s
FAIL
```